### PR TITLE
feat: (CXF 90011) Updating Metal & Fabric Provider with NIMF functionality

### PR DESCRIPTION
**Pull request by srushti-patl**
- Added authorization_code to metal_interconnection schema.
- Added buildSharedPortVCVLANCreateRequest() function for the connection type shared_port_vlan request.
- Added DataSource and Resource Acc Tests for Vlan Port Share functionality.
- Added "METAL_NETWORK" to Fabric connection Access Point Type schema
- Updated Fabric Go SDK Version to 0.9.0

### CODE REVIEW

File: internal/resources/metal/connection/resource_schema.go [hunk 1]
```diff
@@ -82,6 +82,7 @@ func resourceSchema(ctx context.Context) schema.Schema {
 					stringvalidator.OneOf(
 						string(metalv1.INTERCONNECTIONTYPE_DEDICATED),
 						string(metalv1.INTERCONNECTIONTYPE_SHARED),
+						string(metalv1.INTERCONNECTIONTYPE_SHARED_PORT_VLAN),
```
Comment (ctreatma) [inline internal/resources/metal/connection/resource_schema.go:85]: The `Description` of this attribute a few lines above should be updated to reflect the new supported value.

File: internal/resources/metal/connection/resource_schema.go [hunk 2]
```diff
@@ -194,6 +195,10 @@ func resourceSchema(ctx context.Context) schema.Schema {
 					listplanmodifier.UseStateForUnknown(),
 				},
 			},
+			"authorization_code": schema.StringAttribute{
+				Description: "Only used with Fabric Shared connection. Fabric uses this token to be able to give more detailed information about the Metal end of the network, when viewing resources from within Fabric.",
```
Comment (ctreatma) [inline internal/resources/metal/connection/resource_schema.go:199]: Does this work with both `shared` and `shared_port_vlan`?  What are the customer-facing terms for `dedicated`, `shared`, and `shared_port_vlan` connections, and do we define that mapping somewhere in Equinix documentation?

Comment (srushti-patl) [reply internal/resources/metal/connection/resource_schema.go:199]: Yes, the authorization_code is present in both request responses, so we can use it for both requests. 
'dedicated', 'shared', and 'shared_port_vlan' are the customer-facing terms. I have referenced this Metal API documentation page.
https://deploy.equinix.com/developers/api/metal#tag/Interconnections/operation/createProjectInterconnection

**Review by adammohammed — COMMENTED** [review]
Will there be a related PR for the `Metal2SpAws` create body on the Fabric /connections API?

File: internal/resources/metal/connection/resource_test.go [hunk 3]
```diff
@@ -273,6 +298,34 @@ func TestAccMetalConnection_sharedVlan(t *testing.T) {
...
+		ProtoV5ProviderFactories: acceptance.ProtoV5ProviderFactories,
+		CheckDestroy:             testAccMetalConnectionCheckDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccMetalConnectionConfig_SharedPort(rs),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"equinix_metal_connection.test", "metro", "sv"),
+					resource.TestCheckResourceAttr(
+						"equinix_metal_connection.test", "contact_email", "tfacc@example.com"),
```
[the remark this answers is not in the document]
Comment (srushti-patl) [reply internal/resources/metal/connection/resource_test.go:316]: I have added authorization_code parameter check in acc tests

**Comment by srushti-patl** [comment]
> Will there be a related PR for the `Metal2SpAws` create body on the Fabric /connections API?

We have included the required Fabric Provider changes in this PR only. There is only one change required in the Fabric Provider to support NIMF functionality.
This is the PR for Metal 2 AWS example: https://github.com/equinix/terraform-equinix-fabric/pull/34/files

**Review by thogarty — APPROVED** [review]
LGTM for Fabric changes. Metal side looks good as well but leaving that for those who have more familiarity with the product side of things.

**Review by 0xch4z — APPROVED** [review]

**Comment by 0xch4z** [comment]
might need a rebase, but LGTM

File: go.mod [hunk 4]
```diff
@@ -4,7 +4,7 @@ go 1.21
 
 require (
 	github.com/antihax/optional v1.0.0
-	github.com/equinix-labs/fabric-go v0.7.1
+	github.com/equinix-labs/fabric-go v0.8.0
```
Comment (ctreatma) [inline go.mod:7]: It looks like the latest is v0.9.0; is there a reason not to upgrade to the latest here?

**Review by ctreatma — COMMENTED** [review]
I had one question about the fabric-go upgrade, and I don't see the changes Adam suggested for testing that the authorization_code attribute has a value.

File: go.mod (hunk 4 above)
Comment (srushti-patl) [reply go.mod:7]: I will upgrade it to v0.9.0

**Comment by srushti-patl** [comment]
> I had one question about the fabric-go upgrade, and I don't see the changes Adam suggested for testing that the authorization_code attribute has a value.



> I had one question about the fabric-go upgrade, and I don't see the changes Adam suggested for testing that the authorization_code attribute has a value.

I have upgraded fabric-go-sdk version to 0.9.0 and I have already added authorization_code parameter check in metal connection resource and data source acc tests with TestCheckResourceAttrSet() function as authorization_code's value is dynamic.

**Comment by adammohammed** [comment]
@ctreatma I see it now, the addition of `METAL_NETWORK` is all that was required on the Fabric side, since it already had the `authentication_key` field from other request bodies.

**Review by adammohammed — APPROVED** [review]
LGTM, though there are some failing acceptance tests

**Comment by srushti-patl** [comment]
Screenshots of resource and data source Acceptance tests passing locally.
The Metal Acceptance test case workflow failing for these two test cases because the user is not whitelisted for NIMF functionality.

Screenshot 2024-04-01 at 10 50 04 AM
Screenshot 2024-04-01 at 10 54 30 AM

**Comment by ctreatma** [comment]
@adammohammed how can we get the CI user whitelisted for NIMF?

**Comment by adammohammed** [comment]
@ctreatma on it

**Review by ctreatma — APPROVED** [review]
I see the new tests passing in [run #2 of the Metal acceptance tests](https://github.com/equinix/terraform-provider-equinix/actions/runs/8485971049/job/23315311687?pr=626).  Test failures are all in unrelated tests due to platform capacity ("no available servers in metro"), or tests using recently-disabled features ("facility is not supported"), or tests that are known to flake (spot market tests).

**Pull request merged**